### PR TITLE
Add a warning comment on commits that violate https://github.com/NixO…

### PR DESCRIPTION
**Pull request by domenkozar**
On master and release branches, this action checks if the commit pushed belongs to at least one PR that has the same commit set as a merge commit.

If it doesn't, it comments on the commit about https://github.com/NixOS/nixpkgs/issues/118661

### CODE REVIEW

File: .github/workflows/direct-push.yml
Comment (matthiasbeyer) [inline .github/workflows/direct-push.yml:19]: Possibly a typo here? Should this say "Warn that it was a direct push"?

File: .github/workflows/direct-push.yml
Comment (alyssais) [inline .github/workflows/direct-push.yml:24]:
```suggestion
          @${{ github.actor }}, you pushed a commit directly to master/release branch
```

Otherwise I think it sounds a little passive aggressive.

**Comment by sternenseemann** [comment]
I'm a bit sceptical whether it is good UX to automatically flag commits that are still deemed acceptable (e. g. evaluation fixes which don't benefit from going through a PR).

**Comment by alyssais** [comment]
Just a note in case it confuses anybody else: looking at the code, it wasn't clear to me whether this would work for squash/rebase merges, but I tested the curl command with a recent rebase merge, and it worked fine. :)

**Comment by alyssais** [comment]
o_O that wasn't supposed to happen.  Perhaps that action needs some tuning...

**Comment by domenkozar** [comment]
cc @zowoq per https://github.com/NixOS/nixpkgs/commit/fc3b385a4f1ec284f57fd8539eb90b7afb96add5#diff-63e050857aa8769f33e42aef9b82d1516c410dca42b52632f4e2d395a86abf92

**Comment by Synthetica9** [comment]
> I'm a bit sceptical whether it is good UX to automatically flag commits that are still deemed acceptable (e. g. evaluation fixes which don't benefit from going through a PR).

I think it would be good to detect these and not leave a comment (if the file doesn't parse in the parent commit but does parse in the master commit), but getting the MVP with this PR in seems like a good step.

**Review by grahamc — APPROVED** [review]

File: .github/workflows/direct-push.yml
Comment (SuperSandro2000) [inline .github/workflows/direct-push.yml:24]: What happens if I push someone else commit to master? Is it still pining me or the other person?

File: .github/workflows/direct-push.yml
Comment (SuperSandro2000) [inline .github/workflows/direct-push.yml:28]:
```suggestion
          on [https://github.com/NixOS/nixpkgs/issues/118661](https://github.com/NixOS/nixpkgs/issues/118661).
```

Not sure if GitHub would linkify that automatically and I can't really test it right now.

Comment (Synthetica9) [reply .github/workflows/direct-push.yml:28]: https://github.com/NixOS/nixpkgs/issues/118661. I assume github has one markdown parser that handles all markdown so if it works here it should work everywhere?

Comment (SuperSandro2000) [reply .github/workflows/direct-push.yml:28]: Makes sense. I was just not sure if Refined GitHub does it.

File: .github/workflows/direct-push.yml
Comment (domenkozar) [reply .github/workflows/direct-push.yml:24]: It will ping the pusher.

**Comment by domenkozar** [comment]
> > I'm a bit sceptical whether it is good UX to automatically flag commits that are still deemed acceptable (e. g. evaluation fixes which don't benefit from going through a PR).
> 
> I think it would be good to detect these and not leave a comment (if the file doesn't parse in the parent commit but does parse in the master commit), but getting the MVP with this PR in seems like a good step.

Those can be skipped via `[skip ci]` prefix in the commit message, but we can add something else to escape it too like "[direct push]" or so. 

Either way, we'll see what kind of commits are really needed.

**Comment by domenkozar** [comment]
I'm planning to merge this by the end of the day unless something comes up as a blocker.

**Comment by 7c6f434c** [comment]
> Either way, we'll see what kind of commits are really needed.

With this approach, maybe start with a milder wording (like the one I proposed in the other copy of the discussion)? I guess as a message-only change it doesn't need any extra attention to doing correct/incorrect thing in the first place.

File: .github/workflows/direct-push.yml
Comment (Pamplemousse) [inline .github/workflows/direct-push.yml:29]: The link is broken on GH on my side, as it counts the trailing `.` as part of the URL.

**Comment by domenkozar** [comment]
> > Either way, we'll see what kind of commits are really needed.
> 
> With this approach, maybe start with a milder wording (like the one I proposed in the other copy of the discussion)? I guess as a message-only change it doesn't need any extra attention to doing correct/incorrect thing in the first place.

I've remassaged the issue description and the commit comment is written in an informative style.
I still think the message should be clear, as most committers will want to know what action to take.

We can bikeshed the wording after this is in master - feel free to make a PR!

**Comment by vcunat** [comment]
I've experienced some false positives, but only on a very small fraction of my pushes.  I don't know how they happen.  And if they're rare enough, I don't consider them being a problem &ndash; it's just a notification that informs you.  Perhaps the text could mention that it's not 100% accurate though.

**Comment by domenkozar** [comment]
@timokau it's a known issue, I'd like to add a small delay as an attempt to fix this.

**Comment by timokau** [comment]
Alright, that's good to know. I think mentioning the false-positives in the warning would be helpful if we cannot reliably fix this. It might avoid some confusion.

**Pull request merged**
